File: cc_sessions/python/parallel_orchestrator.py
#!/usr/bin/env python3
"""
Parallel Agent Orchestrator

Manages execution of multiple Task agents simultaneously to parallelize work.
"""

import json
import subprocess
import time
from pathlib import Path
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass, field
from enum import Enum

# Import from hooks
from cc_sessions.python.hooks.shared_state import PROJECT_ROOT


class AgentStatus(str, Enum):
    """Status of a parallel agent task"""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class AgentTask:
    """Represents a single agent task in parallel execution"""
    id: str
    agent_name: str
    prompt: str
    status: AgentStatus = AgentStatus.PENDING
    process_id: Optional[int] = None
    output: str = ""
    error: str = ""
    exit_code: Optional[int] = None
    start_time: Optional[float] = None
    end_time: Optional[float] = None

    @property
    def duration(self) -> Optional[float]:
        """Calculate task duration in seconds"""
        if self.start_time and self.end_time:
            return self.end_time - self.start_time
        return None

    def to_dict(self) -> Dict:
        """Convert to dictionary for serialization"""
        return {
            'id': self.id,
            'agent_name': self.agent_name,
            'prompt': self.prompt,
            'status': self.status.value,
            'process_id': self.process_id,
            'output': self.output,
            'error': self.error,
            'exit_code': self.exit_code,
            'start_time': self.start_time,
            'end_time': self.end_time,
            'duration': self.duration
        }


@dataclass
class OrchestrationResult:
    """Result of parallel orchestration execution"""
    total_tasks: int
    completed: int
    failed: int
    total_duration: float
    tasks: List[AgentTask] = field(default_factory=list)

    def to_dict(self) -> Dict:
        """Convert to dictionary for serialization"""
        return {
            'total_tasks': self.total_tasks,
            'completed': self.completed,
            'failed': self.failed,
            'total_duration': self.total_duration,
            'tasks': [task.to_dict() for task in self.tasks]
        }

    def __str__(self) -> str:
        """Human-readable summary"""
        lines = [
            f"\n=== Parallel Orchestration Results ===",
            f"Total Tasks: {self.total_tasks}",
            f"Completed: {self.completed}",
            f"Failed: {self.failed}",
            f"Total Duration: {self.total_duration:.2f}s",
            f"\nIndividual Task Results:"
        ]

        for task in self.tasks:
            status_symbol = "✓" if task.status == AgentStatus.COMPLETED else "✗"
            duration_str = f"{task.duration:.2f}s" if task.duration else "N/A"
            lines.append(f"  {status_symbol} {task.agent_name} ({task.id}): {task.status.value} - {duration_str}")

        return "\n".join(lines)


class ParallelAgentOrchestrator:
    """
    Orchestrates multiple Task agents to run in parallel.

    This allows complex operations like code review to be split across
    multiple agents analyzing different files/modules simultaneously.
    """

    def __init__(self, max_concurrent: int = 4):
        """
        Initialize the orchestrator.

        Args:
            max_concurrent: Maximum number of agents to run simultaneously
        """
        self.max_concurrent = max_concurrent
        self.tasks: List[AgentTask] = []
        self.running_tasks: List[AgentTask] = []

    def add_task(self, agent_name: str, prompt: str, task_id: Optional[str] = None) -> AgentTask:
        """
        Add a task to the orchestration queue.

        Args:
            agent_name: Name of the agent to invoke (e.g., 'code-review', 'context-gathering')
            prompt: The prompt to send to the agent
            task_id: Optional custom ID for the task

        Returns:
            The created AgentTask
        """
        if not task_id:
            task_id = f"{agent_name}_{len(self.tasks) + 1}"

        task = AgentTask(
            id=task_id,
            agent_name=agent_name,
            prompt=prompt
        )
        self.tasks.append(task)
        return task

    def _start_agent_process(self, task: AgentTask) -> subprocess.Popen:
        """
        Start a Task agent process for the given task.

        Args:
            task: The task to execute

        Returns:
            The subprocess.Popen object
        """
        # TODO: This needs to invoke the Claude Code Task tool
        # For now, this is a placeholder that would need integration with Claude Code's API
        # In practice, this would call out to Claude Code's agent execution system

        # Placeholder command - would actually invoke Task tool through Claude Code
        command = [
            'echo',
            f"[PLACEHOLDER] Running agent {task.agent_name} with prompt: {task.prompt[:50]}..."
        ]

        process = subprocess.Popen(
            command,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )

        task.process_id = process.pid
        task.status = AgentStatus.RUNNING
        task.start_time = time.time()

        return process

    def _poll_running_tasks(self) -> None:
        """Check status of running tasks and move completed ones."""
        completed = []

        for task in self.running_tasks:
            # In real implementation, would check actual process status
            # For now, simulate completion
            task.status = AgentStatus.COMPLETED
            task.end_time = time.time()
            task.exit_code = 0
            completed.append(task)

        # Remove completed tasks from running list
        for task in completed:
            self.running_tasks.remove(task)

    def execute_all(self, timeout: Optional[float] = None) -> OrchestrationResult:
        """
        Execute all queued tasks in parallel.

        Args:
            timeout: Optional timeout in seconds for the entire orchestration

        Returns:
            OrchestrationResult with execution summary
        """
        start_time = time.time()
        pending_tasks = [t for t in self.tasks if t.status == AgentStatus.PENDING]

        while pending_tasks or self.running_tasks:
            # Check for timeouts
            if timeout and (time.time() - start_time) > timeout:
                # Mark all pending/running tasks as failed
                for task in pending_tasks + self.running_tasks:
                    if task.status in (AgentStatus.PENDING, AgentStatus.RUNNING):
                        task.status = AgentStatus.FAILED
                        task.error = "Timeout exceeded"
                        task.end_time = time.time()
                break

            # Start new tasks up to max_concurrent limit
            while len(self.running_tasks) < self.max_concurrent and pending_tasks:
                task = pending_tasks.pop(0)
                try:
                    self._start_agent_process(task)
                    self.running_tasks.append(task)
                except Exception as e:
                    task.status = AgentStatus.FAILED
                    task.error = str(e)
                    task.end_time = time.time()

            # Poll running tasks for completion
            self._poll_running_tasks()

            # Small sleep to avoid busy-waiting
            if self.running_tasks:
                time.sleep(0.1)

        # Calculate results
        total_duration = time.time() - start_time
        completed = sum(1 for t in self.tasks if t.status == AgentStatus.COMPLETED)
        failed = sum(1 for t in self.tasks if t.status == AgentStatus.FAILED)

        return OrchestrationResult(
            total_tasks=len(self.tasks),
            completed=completed,
            failed=failed,
            total_duration=total_duration,
            tasks=self.tasks.copy()
        )

    def execute_and_wait(self, task_definitions: List[Tuple[str, str]]) -> OrchestrationResult:
        """
        Convenience method to add tasks and execute them immediately.

        Args:
            task_definitions: List of (agent_name, prompt) tuples

        Returns:
            OrchestrationResult with execution summary

        Example:
            orchestrator = ParallelAgentOrchestrator(max_concurrent=3)
            result = orchestrator.execute_and_wait([
                ('code-review', 'Review src/auth/'),
                ('code-review', 'Review src/api/'),
                ('code-review', 'Review src/database/')
            ])
        """
        for agent_name, prompt in task_definitions:
            self.add_task(agent_name, prompt)

        return self.execute_all()


def parallel_code_review(file_paths: List[str], max_concurrent: int = 3) -> OrchestrationResult:
    """
    Perform parallel code review across multiple files/directories.

    Args:
        file_paths: List of file or directory paths to review
        max_concurrent: Maximum number of reviews to run in parallel

    Returns:
        OrchestrationResult with all review results

    Example:
        result = parallel_code_review([
            'src/auth/',
            'src/api/',
            'src/database/'
        ])
        print(result)
    """
    orchestrator = ParallelAgentOrchestrator(max_concurrent=max_concurrent)

    for path in file_paths:
        prompt = f"Perform a thorough code review of {path}. Focus on security, performance, and code quality."
        orchestrator.add_task('code-review', prompt, task_id=f"review_{Path(path).name}")

    return orchestrator.execute_all()


def parallel_context_gathering(task_files: List[str], max_concurrent: int = 3) -> OrchestrationResult:
    """
    Gather context for multiple tasks in parallel.

    Args:
        task_files: List of task file paths
        max_concurrent: Maximum number of context gatherings to run in parallel

    Returns:
        OrchestrationResult with all context gathering results

    Example:
        result = parallel_context_gathering([
            'sessions/tasks/h-implement-auth.md',
            'sessions/tasks/m-add-logging.md'
        ])
    """
    orchestrator = ParallelAgentOrchestrator(max_concurrent=max_concurrent)

    for task_file in task_files:
        task_path = Path(task_file)
        prompt = f"Create a comprehensive context manifest for the task in {task_file}"
        orchestrator.add_task('context-gathering', prompt, task_id=f"context_{task_path.stem}")

    return orchestrator.execute_all()


# Example usage and testing
if __name__ == "__main__":
    # Example: Parallel code review
    print("=== Example: Parallel Code Review ===")
    result = parallel_code_review([
        'cc_sessions/python/hooks/',
        'cc_sessions/python/api/',
        'cc_sessions/protocols/'
    ], max_concurrent=2)
    print(result)

    # Example: Custom orchestration
    print("\n=== Example: Custom Orchestration ===")
    orchestrator = ParallelAgentOrchestrator(max_concurrent=3)
    orchestrator.add_task('code-review', 'Review authentication module')
    orchestrator.add_task('code-review', 'Review API endpoints')
    orchestrator.add_task('logging', 'Consolidate task logs')

    result = orchestrator.execute_all()
    print(result)
    print(f"\nJSON output:\n{json.dumps(result.to_dict(), indent=2)}")

File: cc_sessions/javascript/parallel_orchestrator.js
#!/usr/bin/env node
/**
 * Parallel Agent Orchestrator
 *
 * Manages execution of multiple Task agents simultaneously to parallelize work.
 */

const { spawn } = require('child_process');
const path = require('path');

// Import from hooks
const { PROJECT_ROOT } = require('./hooks/shared_state.js');

/**
 * Status of a parallel agent task
 */
class AgentStatus {
    static PENDING = 'pending';
    static RUNNING = 'running';
    static COMPLETED = 'completed';
    static FAILED = 'failed';
}

/**
 * Represents a single agent task in parallel execution
 */
class AgentTask {
    /**
     * @param {Object} params
     * @param {string} params.id - Task identifier
     * @param {string} params.agentName - Name of the agent
     * @param {string} params.prompt - Prompt to send to the agent
     * @param {string} [params.status] - Current status
     * @param {number} [params.processId] - Process ID
     * @param {string} [params.output] - Output from the task
     * @param {string} [params.error] - Error from the task
     * @param {number} [params.exitCode] - Exit code
     * @param {number} [params.startTime] - Start timestamp
     * @param {number} [params.endTime] - End timestamp
     */
    constructor({
        id,
        agentName,
        prompt,
        status = AgentStatus.PENDING,
        processId = null,
        output = '',
        error = '',
        exitCode = null,
        startTime = null,
        endTime = null
    }) {
        this.id = id;
        this.agentName = agentName;
        this.prompt = prompt;
        this.status = status;
        this.processId = processId;
        this.output = output;
        this.error = error;
        this.exitCode = exitCode;
        this.startTime = startTime;
        this.endTime = endTime;
    }

    /**
     * Calculate task duration in seconds
     * @returns {number|null}
     */
    get duration() {
        if (this.startTime !== null && this.endTime !== null) {
            return this.endTime - this.startTime;
        }
        return null;
    }

    /**
     * Convert to dictionary for serialization
     * @returns {Object}
     */
    toDict() {
        return {
            id: this.id,
            agentName: this.agentName,
            prompt: this.prompt,
            status: this.status,
            processId: this.processId,
            output: this.output,
            error: this.error,
            exitCode: this.exitCode,
            startTime: this.startTime,
            endTime: this.endTime,
            duration: this.duration
        };
    }
}

/**
 * Result of parallel orchestration execution
 */
class OrchestrationResult {
    /**
     * @param {Object} params
     * @param {number} params.totalTasks - Total number of tasks
     * @param {number} params.completed - Number of completed tasks
     * @param {number} params.failed - Number of failed tasks
     * @param {number} params.totalDuration - Total execution time
     * @param {AgentTask[]} [params.tasks] - List of tasks
     */
    constructor({
        totalTasks,
        completed,
        failed,
        totalDuration,
        tasks = []
    }) {
        this.totalTasks = totalTasks;
        this.completed = completed;
        this.failed = failed;
        this.totalDuration = totalDuration;
        this.tasks = tasks;
    }

    /**
     * Convert to dictionary for serialization
     * @returns {Object}
     */
    toDict() {
        return {
            totalTasks: this.totalTasks,
            completed: this.completed,
            failed: this.failed,
            totalDuration: this.totalDuration,
            tasks: this.tasks.map(task => task.toDict())
        };
    }

    /**
     * Human-readable summary
     * @returns {string}
     */
    toString() {
        const lines = [
            '\n=== Parallel Orchestration Results ===',
            `Total Tasks: ${this.totalTasks}`,
            `Completed: ${this.completed}`,
            `Failed: ${this.failed}`,
            `Total Duration: ${this.totalDuration.toFixed(2)}s`,
            '\nIndividual Task Results:'
        ];

        for (const task of this.tasks) {
            const statusSymbol = task.status === AgentStatus.COMPLETED ? '✓' : '✗';
            const durationStr = task.duration !== null ? `${task.duration.toFixed(2)}s` : 'N/A';
            lines.push(`  ${statusSymbol} ${task.agentName} (${task.id}): ${task.status} - ${durationStr}`);
        }

        return lines.join('\n');
    }
}

/**
 * Orchestrates multiple Task agents to run in parallel.
 *
 * This allows complex operations like code review to be split across
 * multiple agents analyzing different files/modules simultaneously.
 */
class ParallelAgentOrchestrator {
    /**
     * Initialize the orchestrator.
     *
     * @param {number} [maxConcurrent=4] - Maximum number of agents to run simultaneously
     */
    constructor(maxConcurrent = 4) {
        this.maxConcurrent = maxConcurrent;
        this.tasks = [];
        this.runningTasks = [];
    }

    /**
     * Add a task to the orchestration queue.
     *
     * @param {string} agentName - Name of the agent to invoke (e.g., 'code-review', 'context-gathering')
     * @param {string} prompt - The prompt to send to the agent
     * @param {string} [taskId] - Optional custom ID for the task
     * @returns {AgentTask} The created AgentTask
     */
    addTask(agentName, prompt, taskId = null) {
        if (!taskId) {
            taskId = `${agentName}_${this.tasks.length + 1}`;
        }

        const task = new AgentTask({
            id: taskId,
            agentName,
            prompt
        });
        this.tasks.push(task);
        return task;
    }

    /**
     * Start a Task agent process for the given task.
     *
     * @param {AgentTask} task - The task to execute
     * @returns {Promise<void>}
     * @private
     */
    async _startAgentProcess(task) {
        // TODO: This needs to invoke the Claude Code Task tool
        // For now, this is a placeholder that would need integration with Claude Code's API
        // In practice, this would call out to Claude Code's agent execution system

        // Placeholder command - would actually invoke Task tool through Claude Code
        const process = spawn('echo', [
            `[PLACEHOLDER] Running agent ${task.agentName} with prompt: ${task.prompt.slice(0, 50)}...`
        ]);

        task.processId = process.pid;
        task.status = AgentStatus.RUNNING;
        task.startTime = Date.now() / 1000; // Convert to seconds

        // Collect output
        let stdout = '';
        let stderr = '';

        process.stdout.on('data', (data) => {
            stdout += data.toString();
        });

        process.stderr.on('data', (data) => {
            stderr += data.toString();
        });

        // Store the process for later polling
        task._process = process;
        task._stdout = stdout;
        task._stderr = stderr;
    }

    /**
     * Check status of running tasks and move completed ones.
     * @private
     */
    _pollRunningTasks() {
        const completed = [];

        for (const task of this.runningTasks) {
            const process = task._process;

            // Check if process has exited
            if (process && process.exitCode !== null) {
                task.status = process.exitCode === 0 ? AgentStatus.COMPLETED : AgentStatus.FAILED;
                task.endTime = Date.now() / 1000;
                task.exitCode = process.exitCode;
                task.output = task._stdout || '';
                task.error = task._stderr || '';
                completed.push(task);

                // Clean up process references
                delete task._process;
                delete task._stdout;
                delete task._stderr;
            } else if (!process) {
                // Simulate completion for placeholder implementation
                task.status = AgentStatus.COMPLETED;
                task.endTime = Date.now() / 1000;
                task.exitCode = 0;
                completed.push(task);
            }
        }

        // Remove completed tasks from running list
        for (const task of completed) {
            const index = this.runningTasks.indexOf(task);
            if (index !== -1) {
                this.runningTasks.splice(index, 1);
            }
        }
    }

    /**
     * Execute all queued tasks in parallel.
     *
     * @param {number} [timeout] - Optional timeout in seconds for the entire orchestration
     * @returns {Promise<OrchestrationResult>} Execution summary
     */
    async executeAll(timeout = null) {
        const startTime = Date.now() / 1000;
        const pendingTasks = this.tasks.filter(t => t.status === AgentStatus.PENDING);

        while (pendingTasks.length > 0 || this.runningTasks.length > 0) {
            // Check for timeouts
            if (timeout && (Date.now() / 1000 - startTime) > timeout) {
                // Mark all pending/running tasks as failed
                for (const task of [...pendingTasks, ...this.runningTasks]) {
                    if (task.status === AgentStatus.PENDING || task.status === AgentStatus.RUNNING) {
                        task.status = AgentStatus.FAILED;
                        task.error = 'Timeout exceeded';
                        task.endTime = Date.now() / 1000;
                    }
                }
                break;
            }

            // Start new tasks up to max_concurrent limit
            while (this.runningTasks.length < this.maxConcurrent && pendingTasks.length > 0) {
                const task = pendingTasks.shift();
                try {
                    await this._startAgentProcess(task);
                    this.runningTasks.push(task);
                } catch (error) {
                    task.status = AgentStatus.FAILED;
                    task.error = error.message;
                    task.endTime = Date.now() / 1000;
                }
            }

            // Poll running tasks for completion
            this._pollRunningTasks();

            // Small sleep to avoid busy-waiting
            if (this.runningTasks.length > 0) {
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        }

        // Calculate results
        const totalDuration = Date.now() / 1000 - startTime;
        const completed = this.tasks.filter(t => t.status === AgentStatus.COMPLETED).length;
        const failed = this.tasks.filter(t => t.status === AgentStatus.FAILED).length;

        return new OrchestrationResult({
            totalTasks: this.tasks.length,
            completed,
            failed,
            totalDuration,
            tasks: [...this.tasks]
        });
    }

    /**
     * Convenience method to add tasks and execute them immediately.
     *
     * @param {Array<[string, string]>} taskDefinitions - Array of [agentName, prompt] tuples
     * @returns {Promise<OrchestrationResult>} Execution summary
     *
     * @example
     * const orchestrator = new ParallelAgentOrchestrator(3);
     * const result = await orchestrator.executeAndWait([
     *     ['code-review', 'Review src/auth/'],
     *     ['code-review', 'Review src/api/'],
     *     ['code-review', 'Review src/database/']
     * ]);
     */
    async executeAndWait(taskDefinitions) {
        for (const [agentName, prompt] of taskDefinitions) {
            this.addTask(agentName, prompt);
        }

        return await this.executeAll();
    }
}

/**
 * Perform parallel code review across multiple files/directories.
 *
 * @param {string[]} filePaths - List of file or directory paths to review
 * @param {number} [maxConcurrent=3] - Maximum number of reviews to run in parallel
 * @returns {Promise<OrchestrationResult>} All review results
 *
 * @example
 * const result = await parallelCodeReview([
 *     'src/auth/',
 *     'src/api/',
 *     'src/database/'
 * ]);
 * console.log(result.toString());
 */
async function parallelCodeReview(filePaths, maxConcurrent = 3) {
    const orchestrator = new ParallelAgentOrchestrator(maxConcurrent);

    for (const filePath of filePaths) {
        const prompt = `Perform a thorough code review of ${filePath}. Focus on security, performance, and code quality.`;
        const fileName = path.basename(filePath);
        orchestrator.addTask('code-review', prompt, `review_${fileName}`);
    }

    return await orchestrator.executeAll();
}

/**
 * Gather context for multiple tasks in parallel.
 *
 * @param {string[]} taskFiles - List of task file paths
 * @param {number} [maxConcurrent=3] - Maximum number of context gatherings to run in parallel
 * @returns {Promise<OrchestrationResult>} All context gathering results
 *
 * @example
 * const result = await parallelContextGathering([
 *     'sessions/tasks/h-implement-auth.md',
 *     'sessions/tasks/m-add-logging.md'
 * ]);
 */
async function parallelContextGathering(taskFiles, maxConcurrent = 3) {
    const orchestrator = new ParallelAgentOrchestrator(maxConcurrent);

    for (const taskFile of taskFiles) {
        const taskPath = path.parse(taskFile);
        const prompt = `Create a comprehensive context manifest for the task in ${taskFile}`;
        orchestrator.addTask('context-gathering', prompt, `context_${taskPath.name}`);
    }

    return await orchestrator.executeAll();
}

// Example usage and testing
async function main() {
    // Example: Parallel code review
    console.log('=== Example: Parallel Code Review ===');
    const result1 = await parallelCodeReview([
        'cc_sessions/javascript/hooks/',
        'cc_sessions/javascript/api/',
        'cc_sessions/protocols/'
    ], 2);
    console.log(result1.toString());

    // Example: Custom orchestration
    console.log('\n=== Example: Custom Orchestration ===');
    const orchestrator = new ParallelAgentOrchestrator(3);
    orchestrator.addTask('code-review', 'Review authentication module');
    orchestrator.addTask('code-review', 'Review API endpoints');
    orchestrator.addTask('logging', 'Consolidate task logs');

    const result2 = await orchestrator.executeAll();
    console.log(result2.toString());
    console.log(`\nJSON output:\n${JSON.stringify(result2.toDict(), null, 2)}`);
}

// Run examples if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

// Exports
module.exports = {
    AgentStatus,
    AgentTask,
    OrchestrationResult,
    ParallelAgentOrchestrator,
    parallelCodeReview,
    parallelContextGathering
};
